app/DTO/ArrayableDTO.php

```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace App\DTO;
+
+use App\Exceptions\Internal\LycheeLogicException;
+use Illuminate\Contracts\Support\Arrayable;
+use ReflectionClass;
+use ReflectionProperty;
+use function Safe\sprintf;
+
+/**
+ * In some cases, when a DTO does not need to apply casts on attributes
+ * we can directly make use of a reflection which returns an array containing
+ * all the PUBLIC attributes of the DTO.
+ */
+class ArrayableDTO extends DTO
+{
+	/**
+	 * By default, we return an array containing the PUBLIC attributes of the DTO.
+	 *
+	 * @return array the serialized properties of this object
+	 */
+	public function toArray(): array
+	{
+		$result = [];
+		$cls = new ReflectionClass($this);
+		$props = $cls->getProperties(ReflectionProperty::IS_PUBLIC);
+		foreach ($props as $prop) {
+			$value = $prop->getValue($this);
+			if (is_object($value)) {
+				if ($value instanceof Arrayable) {
+					$value = $value->toArray();
+				} else {
+					throw new LycheeLogicException(sprintf("We don't know how to %s.", get_class($value)));
+				}
+			}
```

Comment:
I suspect that we need to consider at least one more case to make this class truly general. But again it is an optional CR and I am not sure if we need it. If the value itself is an array its values might not be primitive values but be objects which need to be converted into arrays as well.

So something along the lines

```php
if (is_array($value)) {
	foreach ($value as &$entry) {
		// we pass $entry by reference to convert objects in place
		if(is_object($entry) {
			if ($entry instanceof Arrayable) {
				$entry = $entry->toArray();
			} else {
				throw new LycheeLogicException('...');
			}
		}
	}
}
```

Reply:
> I suspect that we need to consider at least one more case to make this class truly general.

I did not consider the "_array of object_" because I consider it to be the responsibility of the array provider if an array is passed directly.